pkg/summary/summary.go

```diff
@@ -49,119 +58,144 @@ type Summary struct {
 	Namespaces  []string            `json:"namespaces"`
 }
 
-// Client checks if VPA objects should be created or deleted
-// what? how does it do that? it checks it??
-type Client struct {
-	//changing these two to match the naming in here...but should it be consistent?
-	KubeClient    *kube.ClientInstance
-	KubeClientVPA *kube.VPAClientInstance
-}
+// Summarizer represents a source of generating a summary of VPAs
+type Summarizer struct {
+	Options
 
-var singleton *Client
+	// cached list of vpas
+	vpas []v1beta2.VerticalPodAutoscaler
+}
 
-// GetInstance returns a Client singleton
-func GetInstance() *Client {
-	if singleton == nil {
-		singleton = &Client{
-			KubeClient:    kube.GetInstance(),
-			KubeClientVPA: kube.GetVPAInstance(),
-		}
+// Returns a Summarizer for all goldilocks managed VPAs in all Namespaces
+func NewSummarizer(setters ...Option) *Summarizer {
+	opts := defaultOptions()
+	for _, setter := range setters {
+		setter(opts)
 	}
-	return singleton
-}
 
-// SetInstance sets the singleton using preconstructed k8s and vpa clients. Used for testing.
-func SetInstance(k8s *kube.ClientInstance, vpa *kube.VPAClientInstance) *Client {
-	singleton = &Client{
-		KubeClient:    k8s,
-		KubeClientVPA: vpa,
+	return &Summarizer{
+		Options: *opts,
 	}
-	return singleton
 }
 
-// Run creates a summary of the vpa info for all namespaces.
-func (client *Client) Run(vpaLabels map[string]string, excludeContainers string) (Summary, error) {
-	klog.V(3).Infof("Looking for VPAs with labels: %v", vpaLabels)
-
-	vpaListOptions := metav1.ListOptions{
-		LabelSelector: labels.Set(vpaLabels).String(),
-	}
+// Returns a Summarizer for a known list of VPAs
+func NewSummarizerForVPAs(vpas []v1beta2.VerticalPodAutoscaler, setters ...Option) *Summarizer {
+	summarizer := NewSummarizer(setters...)
 
-	vpas, err := client.KubeClientVPA.Client.AutoscalingV1beta2().VerticalPodAutoscalers("").List(vpaListOptions)
-	if err != nil {
-		klog.Error(err.Error())
-	}
-	klog.V(10).Infof("Found vpas: %v", vpas)
+	// set the cached vpas list directly
+	summarizer.vpas = vpas
 
-	summary, _ := GetInstance().constructSummary(vpas, excludeContainers)
-	return summary, nil
+	return summarizer
 }
 
-func (client *Client) constructSummary(vpas *v1beta2.VerticalPodAutoscalerList, excludeContainers string) (Summary, error) {
+func (s Summarizer) GetSummary() (Summary, error) {
 	var summary Summary
-	if len(vpas.Items) <= 0 {
-		return summary, nil
+	// cached vpas
+	if s.vpas == nil {
+		err := s.UpdateVPAs()
+		if err != nil {
+			return summary, err
+		}
 	}
 
-	containerExclusions := strings.Split(excludeContainers, ",")
-
-	for _, vpa := range vpas.Items {
-		klog.V(8).Infof("Analyzing vpa: %v", vpa.ObjectMeta.Name)
+	if len(s.vpas) <= 0 {
```

Comment:
Does this need to be `<=`? It feels like `== 0` would do the job.

Reply:
Or `< 1`

Reply:
haha this is a habit of mine that I use to make sure that my if statements always cover _all_ cases and not just the cases I think could possibly happen. I can change this if y'all prefer a different style.

Reply:
Nah it's fine as is. It doesn't seem like it could ever be a negative number, but best not to make assumptions 😄